src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -6650,6 +6650,49 @@ struct DynamicGatherOpIsNotDynamic
   }
 };
 
+struct ConcatGather : public OpRewritePattern<stablehlo::ConcatenateOp> {
+  using OpRewritePattern<stablehlo::ConcatenateOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(stablehlo::ConcatenateOp op,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<std::tuple<Attribute, Value, ArrayRef<int64_t>, bool>>
+        gatherParameters;
+    SmallVector<DenseIntElementsAttr> sliceSizesAttrs;
+    for (auto ope : op->getOperands())
+      if (auto operation = llvm::dyn_cast_if_present<stablehlo::GatherOp>(
+              ope.getDefiningOp())) {
+
+        DenseIntElementsAttr sliceSizesAttr;
+        if (!matchPattern(operation.getStartIndices(),
+                          m_Constant(&sliceSizesAttr)))
+          return failure();
+        sliceSizesAttrs.push_back(sliceSizesAttr);
+        gatherParameters.push_back(
+            {operation.getDimensionNumbers(), operation.getOperand(),
+             operation.getSliceSizes(), operation.getIndicesAreSorted()});
```

Comment:
ideally here we should be able to do this if any consecutive ops in the concat are compatible gathers [e.g. it is okay if some ops arent gathers, we'll just concat those segments ourselves into a big gather, then make an outergather where that sequence is replaced with the new gather op